crates/staged-sync/src/test_utils/clique.rs

```diff
@@ -0,0 +1,120 @@
+    pub GethInstance,
+);
+
+impl CliqueGethInstance {
+    /// Sets up a new [`SignerMiddleware`](ethers_middleware::signer_middleware::SignerMiddleware)
+    /// for the [`Geth`](ethers_core::utils::Geth) instance and returns the
+    /// [`CliqueGethInstance`].
+    ///
+    /// The signer is assumed to be the clique signer and the signer for any transactions sent for
+    /// block production.
+    ///
+    /// This also spawns the geth instance.
+    pub async fn new(
+        geth: Geth,
+        signer: Option<SigningKey>,
+    ) -> (Self, SignerMiddleware<Provider<Ws>, Wallet<SigningKey>>) {
+        let signer = signer.unwrap_or_else(|| SigningKey::random(&mut rand::thread_rng()));
+
+        let geth = geth.set_clique_private_key(signer.clone());
+
+        // spawn the geth instance
+        let instance = geth.spawn();
+
+        // create the signer
+        let wallet: LocalWallet = signer.clone().into();
+
+        // set up ethers provider
+        let geth_endpoint = SocketAddr::new([127, 0, 0, 1].into(), instance.port()).to_string();
+        let provider = Provider::<Ws>::connect(format!("ws://{geth_endpoint}")).await.unwrap();
+        let provider =
+            SignerMiddleware::new_with_provider_chain(provider, wallet.clone()).await.unwrap();
+
+        (Self(instance), provider)
+    }
+
+    /// Prints the logs of the [`Geth`](ethers_core::utils::Geth) instance in a new
+    /// [`task`](tokio::task).
+    #[allow(dead_code)]
+    pub async fn print_logs(&mut self) {
+        // take the stderr of the geth instance and print it
+        let stderr = self.0.stderr().unwrap();
+
+        // print logs in a new task
+        let mut err_reader = BufReader::new(stderr);
+
+        tokio::spawn(async move {
+            loop {
+                if let (Ok(line), line_str) = {
+                    let mut buf = String::new();
+                    (err_reader.read_line(&mut buf), buf.clone())
+                } {
+                    if line == 0 {
+                        break
+                    }
+                    if !line_str.is_empty() {
+                        dbg!(line_str);
+                    }
+                }
+            }
+        });
+    }
+
+    /// Prevents the [`Geth`](ethers_core::utils::Geth) instance from blocking due to the `stderr`
+    /// filling up.
+    pub async fn prevent_blocking(&mut self) {
+        // take the stderr of the geth instance and print it
+        let stderr = self.0.stderr().unwrap();
+
+        // print logs in a new task
+        let mut err_reader = BufReader::new(stderr);
+
+        tokio::spawn(async move {
+            loop {
+                if let (Ok(line), _line_str) = {
+                    let mut buf = String::new();
+                    (err_reader.read_line(&mut buf), buf.clone())
+                } {
+                    if line == 0 {
+                        break
+                    }
+                }
+            }
```

Comment:
this is just for debugging, right?

Reply:
`print_logs` is for debugging, but `prevent_blocking` is necessary to prevent the stderr pipe buffer from filling up and causing the test to hang. Using `Drop` doesn't seem to work here because it causes geth to crash (probably due to `EPIPE`, but I'm not entirely sure). I'm not sure if there's a better way to prevent the stderr pipe buffer from filling though

Reply:
we can't do it by default because `Geth` captures the stderr to make sure rpc, p2p, etc are up when `spawn` is called. This means we automatically pipe the `Command`'s output with `Stdio::piped()`. Maybe we could spawn `cat` or something, with geth's stderr as `cat`'s `stdin`, and pipe cat to /dev/null